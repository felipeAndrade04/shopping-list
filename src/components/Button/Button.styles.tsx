import { colors } from '@app/theme';
import { RectButton } from 'react-native-gesture-handler';
import styled from 'styled-components/native';
import { ButtonProps } from './Button.types';

export const Container = styled(RectButton)<ButtonProps>`
  background-color: ${(props) => (props.variant === 'outline' ? colors.white : props.color)};
  border: ${(props) => (props.variant === 'outline' ? `2px solid ${props.color}` : 'none')};
  width: 100%;
  height: 56px;
  padding: 8px 12px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 6px;
  margin-top: ${(props) => props.marginTop || 0}px;
  margin-right: ${(props) => props.marginRight || 0}px;
  margin-bottom: ${(props) => props.marginBottom || 0}px;
  margin-left: ${(props) => props.marginLeft || 0}px;
`;

export const Text = styled.Text<Pick<ButtonProps, 'color' | 'variant'>>`
  font-size: 18px;
  font-weight: 700;
  color: ${(props) => (props.variant === 'outline' ? props.color : colors.white)};
`;
