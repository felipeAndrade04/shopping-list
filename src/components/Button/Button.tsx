import { colors } from '@app/theme';
import React from 'react';
import * as S from './Button.styles';
import { ButtonProps } from './Button.types';
import { ActivityIndicator } from 'react-native';

export function Button({
  children,
  variant = 'solid',
  color = colors.main,
  isLoading,
  enabled = true,
  onPress,
  ...rest
}: ButtonProps) {
  return (
    <S.Container
      color={color}
      variant={variant}
      isLoading={isLoading}
      enabled={enabled}
      onPress={onPress}
      {...rest}
    >
      <S.Text color={color} variant={variant}>
        {isLoading ? (
          <ActivityIndicator
            color={variant === 'outline' ? color : colors.white}
            testID="button-loading"
          />
        ) : (
          children
        )}
      </S.Text>
    </S.Container>
  );
}
