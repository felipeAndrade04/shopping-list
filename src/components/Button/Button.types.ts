import { ReactNode } from 'react';
import { RectButtonProperties } from 'react-native-gesture-handler';

export interface ButtonProps extends RectButtonProperties {
  children: ReactNode | string;
  variant?: 'solid' | 'outline';
  color?: string;
  marginTop?: number;
  marginRight?: number;
  marginBottom?: number;
  marginLeft?: number;
  isLoading?: boolean;
}
