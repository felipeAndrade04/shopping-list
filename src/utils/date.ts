const months = ['jan', 'fev', 'mar', 'abr', 'mai', 'jun', 'jul', 'ago', 'set', 'out', 'nov', 'dez'];

export function formatDate(date: Date) {
  const day = date.getDay();
  const month = date.getMonth();

  return `${day} de ${months[month]}`;
}
