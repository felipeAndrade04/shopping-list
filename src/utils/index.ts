export * from './delay';
export * from './date';
