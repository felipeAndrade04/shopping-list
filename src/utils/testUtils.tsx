import React from 'react';
import { authReducer } from '@app/store';
import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components/native';
import * as theme from '@app/theme';
import { configureStore } from '@reduxjs/toolkit';

let store = configureStore({
  reducer: {
    auth: authReducer,
  },
});

export const wrapper = ({ children }: { children: ReactNode }) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

export function resetStore() {
  store = configureStore({
    reducer: {
      auth: authReducer,
    },
  });
}
