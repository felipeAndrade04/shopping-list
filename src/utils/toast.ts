import EventManager from './EvenManager';

export type ToastType = 'success' | 'info' | 'error';

export interface Toast {
  id?: string;
  type: ToastType;
  text: string;
  duration?: number;
}

export const toastEventManager = new EventManager();

export default function toast({ type, text, duration }: Toast) {
  toastEventManager.emit<Toast>('addtoast', {
    id: new Date().getTime().toString(),
    type,
    text,
    duration,
  });
}
