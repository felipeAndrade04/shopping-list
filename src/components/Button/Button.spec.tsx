import * as React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native';
import { Button } from './Button';
import { wrapper } from '@app/utils/testUtils';

describe('<Button />', () => {
  it('Should be render correctly', () => {
    const tree = render(<Button>Teste</Button>, { wrapper }).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should be render variant outline correctly', () => {
    const tree = render(<Button variant="outline">Teste</Button>, { wrapper }).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should have called the function', () => {
    const mockFn = jest.fn();

    render(<Button onPress={mockFn}>Teste</Button>, { wrapper });

    const button = screen.getByText('Teste');
    fireEvent.press(button);

    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('Should show the loading', () => {
    const buttonText = 'Teste';
    const { queryByTestId, queryByText } = render(<Button isLoading={true}>{buttonText}</Button>, {
      wrapper,
    });

    const activityIndicator = queryByTestId('button-loading');
    const button = queryByText(buttonText);

    expect(activityIndicator).toBeTruthy();
    expect(button).toBeFalsy();
  });
});
