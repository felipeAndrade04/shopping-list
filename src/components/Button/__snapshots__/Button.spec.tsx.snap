// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Should be render correctly 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.105}
  collapsable={false}
  color="#009Cf7"
  delayLongPress={600}
  enabled={true}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onActiveStateChange={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#009Cf7",
      "borderColor": "black",
      "borderRadius": 6,
      "borderStyle": "solid",
      "borderWidth": 0,
      "display": "flex",
      "height": 56,
      "justifyContent": "center",
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
      "opacity": 1,
      "paddingBottom": 8,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 8,
      "width": "100%",
    }
  }
  underlayColor="black"
  variant="solid"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "black",
        "borderBottomLeftRadius": undefined,
        "borderBottomRightRadius": undefined,
        "borderRadius": 6,
        "borderTopLeftRadius": undefined,
        "borderTopRightRadius": undefined,
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <Text
    color="#009Cf7"
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Poppins-Bold",
          "fontSize": 18,
        },
      ]
    }
    variant="solid"
  >
    Teste
  </Text>
</RNGestureHandlerButton>
`;

exports[`<Button /> Should be render variant outline correctly 1`] = `
<RNGestureHandlerButton
  activeOpacity={0.105}
  collapsable={false}
  color="#009Cf7"
  delayLongPress={600}
  enabled={true}
  handlerTag={2}
  handlerType="NativeViewGestureHandler"
  onActiveStateChange={[Function]}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#009Cf7",
      "borderRadius": 6,
      "borderStyle": "solid",
      "borderWidth": 2,
      "display": "flex",
      "height": 56,
      "justifyContent": "center",
      "marginBottom": 0,
      "marginLeft": 0,
      "marginRight": 0,
      "marginTop": 0,
      "opacity": 1,
      "paddingBottom": 8,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 8,
      "width": "100%",
    }
  }
  underlayColor="black"
  variant="outline"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "black",
        "borderBottomLeftRadius": undefined,
        "borderBottomRightRadius": undefined,
        "borderRadius": 6,
        "borderTopLeftRadius": undefined,
        "borderTopRightRadius": undefined,
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <Text
    color="#009Cf7"
    style={
      Array [
        Object {
          "color": "#009Cf7",
          "fontFamily": "Poppins-Bold",
          "fontSize": 18,
        },
      ]
    }
    variant="outline"
  >
    Teste
  </Text>
</RNGestureHandlerButton>
`;
